### PR TITLE
Add WithHeaders option for Zipkin exporter

### PR DESCRIPTION
**Pull request by srijan-27**
- Added `WithHeaders` option func, which allows user to set custom http request headers while exporting spans.
- Closes: #3474

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,7 +41,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Prevent random number generation data-race for experimental rand exemplars in `go.opentelemetry.io/otel/sdk/metric`. (#5456)
 - Fix counting number of dropped attributes of `Record` in `go.opentelemetry.io/otel/sdk/log`. (#5464)
 - Fix panic in baggage creation when a member contains 0x80 char in key or value. (#5494)
-- Correct comments for the priority of the `WithEndpoint` and `WithEndpointURL` options and their coresponding environment variables in  in `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc`. (#5508)
+- Correct comments for the priority of the `WithEndpoint` and `WithEndpointURL` options and their corresponding environment variables in `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc`. (#5508)
```
Comment (dmathieu) [inline CHANGELOG.md:44]: Please keep PRs constrainted to what they're doing. This fix should be another PR.

File: CHANGELOG.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Store and provide the emitted `context.Context` in `ScopeRecords` of `go.opentelemetry.io/otel/sdk/log/logtest`. (#5468)
 - `SimpleProcessor.OnEmit` in `go.opentelemetry.io/otel/sdk/log` no longer allocates a slice which makes it possible to have a zero-allocation log processing using `SimpleProcessor`. (#5493)
 - The `AssertRecordEqual` method to `go.opentelemetry.io/otel/log/logtest` to allow comparison of two log records in tests. (#5499)
+- The `WithHeaders` option to `go.opentelemetry.io/otel/exporters/zipkin` to allow configuring custom http headers while exporting spans. (#3474)
```
Comment (dmathieu) [inline CHANGELOG.md:20]:
```suggestion
- The `WithHeaders` option to `go.opentelemetry.io/otel/exporters/zipkin` to allow configuring custom http headers while exporting spans. (#5530)
```

File: exporters/zipkin/zipkin_test.go [hunk 3]
```diff
@@ -375,3 +375,17 @@ func TestLogrFormatting(t *testing.T) {
 	got := buf.String()
 	assert.Equal(t, want, got)
 }
+
+func TestWithHeaders(t *testing.T) {
```
Comment (dmathieu) [inline exporters/zipkin/zipkin_test.go:379]: Could you add a test checking that the specified headers are indeed added to the outgoing request?

File: CHANGELOG.md (hunk 1 above)
Comment (srijan-27) [reply CHANGELOG.md:44]: This change was done while running `make codespell`, although if not required, will revert this manually.

Comment (dmathieu) [reply CHANGELOG.md:44]: I'm not saying it shouldn't be fixed. Just that it should be in another PR :)

Comment (srijan-27) [reply CHANGELOG.md:44]: cool! have reverted, will push alongwith the tests

Comment (srijan-27) [reply CHANGELOG.md]: @dmathieu - I have made the changes as required. Please check and let me know if there are any suggestions..

File: exporters/zipkin/zipkin_test.go [hunk 4]
```diff
@@ -324,6 +325,29 @@ func TestExportSpans(t *testing.T) {
 	}
 	require.Eventually(t, checkFunc, time.Second, 10*time.Millisecond)
 	require.Equal(t, models, collector.StealModels())
+
+	got := make(chan *http.Request, 1)
```
Comment (dmathieu) [inline exporters/zipkin/zipkin_test.go:329]: This should be a new test, not part of an existing one.

Comment (srijan-27) [reply exporters/zipkin/zipkin_test.go]: Updated the test, please check.

File: exporters/zipkin/zipkin_test.go [hunk 5]
```diff
@@ -375,3 +376,40 @@ func TestLogrFormatting(t *testing.T) {
...
+	defer srv.Close()
+
+	e := &Exporter{
+		url:     srv.URL,
+		client:  srv.Client(),
+		headers: headers,
+	}
+
+	_ = e.ExportSpans(context.Background(), tracetest.SpanStubs{}.Snapshots())
+	req := <-ch
```
Comment (dmathieu) [inline exporters/zipkin/zipkin_test.go:410]: ExportSpans runs synchronously. So once the method has finished running, the request should be populated. Hence, we shouldn't need a channel.

```suggestion
	var req *http.Request
	handler := func(w http.ResponseWriter, r *http.Request) {
		req = r
		w.WriteHeader(http.StatusOK)
	}

	srv := httptest.NewServer(http.HandlerFunc(handler))
	defer srv.Close()

	e := &Exporter{
		url:     srv.URL,
		client:  srv.Client(),
		headers: headers,
	}

	_ = e.ExportSpans(context.Background(), tracetest.SpanStubs{}.Snapshots())
```

Comment (srijan-27) [reply exporters/zipkin/zipkin_test.go:410]: Yes, I tried this way but request comes out to be `nil` as per this. Hence, used channel to read value of request.

Comment (srijan-27) [reply exporters/zipkin/zipkin_test.go:410]: yeah, got it.. I forgot to add a span, will add it.

Comment (srijan-27) [reply exporters/zipkin/zipkin_test.go]: Done the changes, please check.

**Review by dmathieu — APPROVED** [review]

File: exporters/zipkin/zipkin.go [hunk 6]
```diff
@@ -70,6 +72,14 @@ func WithLogr(logger logr.Logger) Option {
 	})
 }
 
+// WithHeaders configures the exporter to use the passed http request headers.
```
Comment (MrAlias) [inline exporters/zipkin/zipkin.go:75]: nit

```suggestion
// WithHeaders configures the exporter to use the passed HTTP request headers.
```
```suggestion
// WithHeaders configures the exporter to use the passed http request headers.
```

File: exporters/zipkin/zipkin.go [hunk 7]
```diff
@@ -132,6 +143,15 @@ func (e *Exporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpa
 		return e.errf("failed to create request to %s: %v", e.url, err)
 	}
 	req.Header.Set("Content-Type", "application/json")
+
+	for k, v := range e.headers {
+		if k == "host" {
```
Comment (MrAlias) [inline exporters/zipkin/zipkin.go:148]: This will not match the valid `"Host"` header.

```suggestion
		if strings.ToLower(k) == "host" {
```

**Review by MrAlias — APPROVED** [review]
